### DISCUSSION TITLE
How to handle unathorized errors thrown by the server

### DISCUSSION THREAD

**Discussion by alinjie** [vercel/next.js#39530] [Help]
Hi,

This issue is kinda hard to explain, so I apologize in advance. 😄 

We're building a fairly large Next.js application and are looking a smooth way to handle unauthorized errors. Our amibition is to throw a custom `UnauthorizedError` if the server response is `401`, check the `err` property from the `getInitialProps` context in `_error.tsx`, return a `statusCode` prop of `401` if `err instanceof UnauthorizedError` evaluates to `true`, and have some conditional logic in `_error.tsx` to render an `Unauthorized` component if `props.statusCode === 401`. If `props.statusCode` is anything else, we want to render some "Internal Server Error" markup instead.

In the [example I've made here](https://github.com/alinjie/next-unauthorized-error-repro) the flow works fine, but in our production application a client side error is thrown at some point after the `Unauthorized` component is rendered, causing the `_error.tsx` page's `getInitialProps` function to re-run on the client side (with a different unknown error), making the `statusCode` prop to equal `500` and causing the `_error.tsx` page to never render the `Unauthorized` component, even though the intitial error was `UnauthorizedError`.

**My question is:**
What is the recommended way, if any, to deal with unauthorized errors thrown by the server? We could of have a `try/catch` in every `getServerSideProps` and redirect if the thrown exception is an instance of `UnauthorizedError`, but this quickly becomes tedious as the application grows. The Auth0 SDK, for example, provides a `withPageAuthRequired` helper where they handle auth stuff, is this the best way to handle this type of logic?

Any input is appreciated.

**Comment by danoc** [comment]
Hey @alinjie, I'm also trying to figure this out. Did you learn anything more about this problem? 

In my case, I'd like to also set the page's status code to 401 rather than 500. Some folks were asking about this in https://github.com/vercel/next.js/issues/11644, but that issue got locked and I can't find another related issue.

**Reply by alinjie** [reply]
Hi!

We ended up mostly fetching data on the client and created a `/error` page where we redirect on client errors. Would've liked to keep fetching on the server and handle the errors in `_error.tsx`, but sadly we didn't find out how to do this efficiently.

Let me know if you find something!

**Comment by b-insh** [comment]
looking for exactly this as well

**Comment by rayrny** [comment]
Hello! I've also been pondering over the same issue and came up with a solution!

The problem we need to address is that when errors occur on the server, they are differentiated with codes like `401, 403, etc`. However, when these errors are delivered to the browser, they are all treated as 500 errors. (Perhaps from the browser's perspective, any error that occurred on the server is considered a 500 error. This is just my assumption!)

So, I came up with an idea to handle this in the `getInitialProps` that runs on the server. I tried to detect the type of error and then overwrite the res object to perform a redirect to a specific page. I haven't tested if it's possible to directly overwrite res with the desired error code like `res.writeHead(401)`.

By using this approach, when a `401 Unauthorized` error occurs on the server, it will redirect to the `{localhost}/401` page, and I can render the desired error page. Also, this idea seems to be extensible in the direction you want.

I'll leave the code I attempted below, in case it might be helpful 😉


```js
// page/_error.js
CustomErrorPage.getInitialProps = ({ res, err }) => {
    const errorName = err.name || err.response?.data.code;
    let statusCode = 500;
    switch (errorName) {
        case 'UnauthorizedError':
            statusCode = 401;
            break;
        default:
            break;
    }

    if (statusCode !== 500) {
        res.writeHead(302, { Location: `/${statusCode}` });
        res.end();
    }

    return { statusCode, myError: statusCode };
};

```
<img width="892" alt="401" src="https://github.com/vercel/next.js/assets/48341341/61896890-6275-4a2f-874e-d757d1f9b6ae">

**Comment by AndonMitev** [comment]
but is it possible to create some kind of interceptor? @rayrny is your solution applicable in nextjs 13 and above?
